feat: Add errcheck linter with golangci-lint

kushthedude opened the pull request:
Refers #1244

kushthedude commented:
/cla signed

kushthedude commented:
/assign @shyamjvs
/cc @wojtek-t

kushthedude commented:
Hey @wojtek-t, I see that the bash script to run golangcilint & lib/gopkg to find modules do not run on macOS due to 
```sh
find: -printf: unknown primary or operator
```

should I also push updated script on the basis of OS_ENV ?

wojtek-t commented on .golangci.yml:
```diff
@@ -1,11 +1,23 @@
     # TODO(oxddr): enable more checkers
     # - deadcode
-    # - errcheck
+    # - misspell
```
I see you're enabling it above - why you added it here?

kushthedude replied:
That is just a configuration being added, I will re-enable it once all the errcheck issues are solved.

wojtek-t commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -70,9 +70,17 @@ func initClusterFlags() {
-	flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	err := flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	if err != nil {
+		fmt.Errorf("cant delete namespaces %v", err)
```
The error is incorrect - it's not about deleting namespace, it's about marking flag as deprecated.

Same below...

kushthedude replied:
Corrected the error message

wojtek-t commented on clusterloader2/pkg/measurement/common/service_creation_latency.go:
```diff
@@ -50,7 +50,11 @@ const (
 
 func init() {
-	measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement)
+	err := measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement)
```
Please look at this:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/etcd_metrics.go#L39

and make it consistent

wojtek-t replied:
Same in other similar places.

kushthedude replied:
fixed for other similar places

wojtek-t commented:
/ok-to-test

kushthedude commented:
Hi @wojtek-t, I am unable to get past the `Fatal call has formatting directive` error in the tests-verify-all, I have already formatted the call by adding `\n` so that go-vet won't fail. Can you please let me know what have I don't wrong here & why the following error is not reported by golangci-lint?

wojtek-t commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -70,9 +70,17 @@ func initClusterFlags() {
-	flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	err := flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	if err != nil {
+		fmt.Errorf("unable to mark flag deprecated %v", err)
```
fmt.Errorf() is just creating error object and not doing anything with it.

you should probably klog.Fatalf() here...
[and in other similar places]

kushthedude replied:
Got it!

wojtek-t commented on clusterloader2/pkg/measurement/common/service_creation_latency.go:
```diff
@@ -50,7 +50,10 @@ const (
 func init() {
-	measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement)
+	if err := measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement); err != nil {
+		fmt.Errorf("cant register service %v", err)
```
Apparently I wasn't enough explicit previously - you should use klog.Fatalf()
e.g. as here:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/profile.go#L46

wojtek-t commented on clusterloader2/pkg/tuningset/global_qps_load.go:
```diff
@@ -59,7 +59,10 @@ func newGlobalQPSLoad(params *api.GlobalQPSLoad) *globalQPSLoad {
-		ql.limiter.Wait(context.TODO())
+		err := ql.limiter.Wait(context.TODO())
+		if err != nil {
+			return
```
This is silently ignoring the error, and skipping all other actions - that's definitely not right.

```
klog.Errorf("Error while waiting for rate limitter: %v - skipping the action", err)
continue
```

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -136,16 +136,22 @@ func cleanup(c *kubernetes.Clientset) {
+		err := c.Core().Services(testNamespace).Delete(
 			svc.GetName(), &metav1.DeleteOptions{})
+		if err != nil {
+			fmt.Println("Failed to get namespace services", err)
```
Failed to get services (i.e. remove namespace)

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -136,16 +136,22 @@ func cleanup(c *kubernetes.Clientset) {
-		c.Core().Namespaces().Create(&api.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}})
+		_, err := c.Core().Namespaces().Create(&api.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}})
+		if err != nil {
+			fmt.Println("Failed to create namespace services", err)
```
remove namespace

wojtek-t commented on perfdash/metrics-downloader-helper.go:
```diff
@@ -48,12 +48,20 @@ func serveHTTPObject(res http.ResponseWriter, req *http.Request, obj interface{}
+		_, err = res.Write([]byte(fmt.Sprintf("<h3>Internal Error</h3><p>%v", err)))
+		if err != nil {
+			klog.Warningf("unable to write error %v", err)
+			return
```
remove this return

wojtek-t commented on perfdash/metrics-downloader-helper.go:
```diff
@@ -48,12 +48,20 @@ func serveHTTPObject(res http.ResponseWriter, req *http.Request, obj interface{}
-		res.Write([]byte(fmt.Sprintf("<h3>Internal Error</h3><p>%v", err)))
+		_, err = res.Write([]byte(fmt.Sprintf("<h3>Internal Error</h3><p>%v", err)))
+		if err != nil {
+			klog.Warningf("unable to write error %v", err)
```
should be Error - not warning

wojtek-t commented on perfdash/metrics-downloader-helper.go:
```diff
@@ -48,12 +48,20 @@ func serveHTTPObject(res http.ResponseWriter, req *http.Request, obj interface{}
-	res.Write(data)
+	_, err = res.Write(data)
+	if err != nil {
+		klog.Warningf("unable to write response data %v", err)
```
should be error not warning

wojtek-t commented on perfdash/metrics-downloader-helper.go:
```diff
@@ -48,12 +48,20 @@ func serveHTTPObject(res http.ResponseWriter, req *http.Request, obj interface{}
+	_, err = res.Write(data)
+	if err != nil {
+		klog.Warningf("unable to write response data %v", err)
+		return
```
remove this return

wojtek-t commented on perfdash/metrics-downloader.go:
```diff
@@ -124,7 +124,7 @@ func (g *Downloader) getJobData(wg *sync.WaitGroup, result JobToCategoryData, re
 					artifacts, err := g.MetricsBkt.ListFilesInBuild(job, buildNumber, searchPrefix)
 					if err != nil || len(artifacts) == 0 {
-						klog.Errorf("Error while looking for %s* in %s build %v: %v", searchPrefix, job, buildNumber, err)
+						klog.Errorf("error while looking for %s* in %s build %v: %v", searchPrefix, job, buildNumber, err)
```
This one was correct - it's not an error, but log, we capitalize those.

kushthedude replied:
Golangci-lint captures this as an error and hence throws error, should I ignore this line for lint check then?

kushthedude replied:
I have changed the error format to just log it so that ci-lint won't throw an error.

wojtek-t commented on perfdash/metrics-downloader.go:
```diff
@@ -133,7 +133,7 @@ func (g *Downloader) getJobData(wg *sync.WaitGroup, result JobToCategoryData, re
 							fmt.Sprintf("artifacts/%v", metricsFileName))
 						if err != nil {
-							klog.Errorf("Error when reading response Body: %v", err)
+							klog.Errorf("error when reading response Body: %v", err)
```
Same here

wojtek-t commented on util-images/probes/pkg/ping/server/server.go:
```diff
@@ -52,5 +52,9 @@ func Run(config *Config) {
+	_, err := w.Write([]byte("pong"))
+	if err != nil {
+		klog.Fatal("unable to write response %v\n", err)
+		return
```
remove return

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -344,14 +350,23 @@ func processCsvData(csvData *string) bool {
-		os.Mkdir(outputFileDirectory, 0766)
+		err := os.Mkdir(outputFileDirectory, 0766)
+		if err != nil {
+			fmt.Printf("Error creating directory", err)
```
Printf requires format, so somethig line:
```
fmt.Printf("Error creating directory: %v", err)
```

Though the patter used in this file seem to be to use Println, so use Println for consistency.

kushthedude replied:
done

wojtek-t commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -70,9 +70,17 @@ func initClusterFlags() {
-	flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	err := flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	if err != nil {
+		klog.Fatalf("unable to mark flag deprecated %v", err)
```
add name of flag here

wojtek-t commented on clusterloader2/cmd/clusterloader.go:
```diff
@@ -70,9 +70,17 @@ func initClusterFlags() {
+	err := flags.MarkDeprecated("delete-stale-namespaces", "specify deleteStaleNamespaces in testconfig file instead.")
+	if err != nil {
+		klog.Fatalf("unable to mark flag deprecated %v", err)
+		return
```
remove return - klog.Fatalf is doing exit 1 anyway.

same below

wojtek-t commented on clusterloader2/pkg/measurement/common/service_creation_latency.go:
```diff
@@ -50,7 +50,10 @@ const (
 func init() {
-	measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement)
+	if err := measurement.Register(serviceCreationLatencyName, createServiceCreationLatencyMeasurement); err != nil {
+		klog.Errorf("cant register service %v", err)
```
Still not what I wanted - please use Fatalf

https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/etcd_metrics.go#L40

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -136,16 +136,22 @@ func cleanup(c *kubernetes.Clientset) {
+		err := c.Core().Services(testNamespace).Delete(
 			svc.GetName(), &metav1.DeleteOptions{})
+		if err != nil {
+			fmt.Println("Failed to get services", err)
```
delete a service

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -136,16 +136,22 @@ func cleanup(c *kubernetes.Clientset) {
-		c.Core().Namespaces().Create(&api.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}})
+		_, err := c.Core().Namespaces().Create(&api.Namespace{ObjectMeta: metav1.ObjectMeta{Name: testNamespace}})
+		if err != nil {
+			fmt.Println("Failed to create services", err)
```
create a service

wojtek-t commented on network/benchmarks/netperf/nptest/nptest.go:
```diff
@@ -509,13 +509,21 @@ func (t *NetPerfRPC) ReceiveOutput(data *WorkerOutput, reply *int) error {
+	err := rpc.Register(baseObject)
+	if err != nil {
+		log.Fatal("failed to register rpc", err)
+		return
```
remove return - Fatal is doing exit 1 underneath

same below

wojtek-t commented on network/benchmarks/netperf/launch.go:
```diff
@@ -344,14 +350,23 @@ func processCsvData(csvData *string) bool {
-		os.Mkdir(outputFileDirectory, 0766)
+		err := os.Mkdir(outputFileDirectory, 0766)
+		if err != nil {
+			fmt.Println("Error creating directory %v", err)
```
remove %v - Println doesn't need it

same below

kushthedude commented:
Resolved all requested changes!

wojtek-t commented on clusterloader2/pkg/measurement/common/slos/pod_startup_latency.go:
```diff
@@ -45,7 +45,10 @@ const (
 func init() {
-	measurement.Register(podStartupLatencyMeasurementName, createPodStartupLatencyMeasurement)
+	if err := measurement.Register(podStartupLatencyMeasurementName, createPodStartupLatencyMeasurement); err != nil {
+		klog.Errorf("cant register service %v", err)
```
klog.Fatalf() as in one of the files above.

wojtek-t commented on util-images/access-tokens/cmd/main.go:
```diff
@@ -62,7 +62,11 @@ func main() {
-	flag.Set("alsologtostderr", "true")
+	err := flag.Set("alsologtostderr", "true")
+	if err != nil {
+		klog.Warningf("Got error while setting flag to true %v", err)
```
add name of the flag

wojtek-t commented on clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus_test.go:
```diff
@@ -46,8 +46,16 @@ var (
-		flag.Set("v", "2")
+		err := flag.Set("logtostderr", "false")
+		if err != nil {
+			klog.Errorf("Unable to set flag %v", err)
```
Instead of this, let's:
- pass `*t.Testing` as a parameter to this function
- use t.Errorf(...) here (and below)

wojtek-t commented on clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus_test.go:
```diff
@@ -565,7 +573,11 @@ func TestLogging(t *testing.T) {
-			gatherer.Gather(executor, time.Now(), config)
+			_, err := gatherer.Gather(executor, time.Now(), config)
+			if err != nil {
+				klog.Errorf("cant gather config %v", err)
```
t.Errorf(...)

wojtek-t commented:
There are two more issues:
benchmarks/netperf/launch.go:367:3: Println call has possible formatting directive %v

--- FAIL: TestLogging (0.00s)
    --- FAIL: TestLogging/print_5_warnings (0.00s)
        api_responsiveness_prometheus_test.go:578: cant gather config top latency metric: there should be no high-latency requests, but: [got: &{Resource:r1 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.2s Count:0 SlowCount:0}; expected perc99 <= 1s]
    --- FAIL: TestLogging/print_all_violations (0.00s)
        api_responsiveness_prometheus_test.go:578: cant gather config top latency metric: there should be no high-latency requests, but: [got: &{Resource:r2 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.9s Count:0 SlowCount:0}; expected perc99 <= 1s got: &{Resource:r3 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.8s Count:0 SlowCount:0}; expected perc99 <= 1s got: &{Resource:r4 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.7s Count:0 SlowCount:0}; expected perc99 <= 1s got: &{Resource:r5 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.6s Count:0 SlowCount:0}; expected perc99 <= 1s got: &{Resource:r6 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.5s Count:0 SlowCount:0}; expected perc99 <= 1s got: &{Resource:r1 Subresource: Verb:POST Scope: Latency:perc50: 0s, perc90: 0s, perc99: 1.2s Count:0 SlowCount:0}; expected perc99 <= 1s]

Modulo fixing those it looks fine

kushthedude commented:
@wojtek-t I fixed the first issue, about the second there is a latency increase of 1.2 sec which can be due to the test parameter I added hence due to which test is failing, how should I go about this?

wojtek-t commented:
> about the second there is a latency increase of 1.2 sec which can be due to the test parameter I added hence due to which test is failing, how should I go about this?

It's not about latency increase - the data are faked in this test and this 1.2s is hardcoded - see line 465.

The real issue is different something is logged differently - you need to take a look into what was logger previously and what is logged now (and why this is different).

wojtek-t commented on clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus_test.go:
```diff
@@ -565,7 +573,11 @@ func TestLogging(t *testing.T) {
-			gatherer.Gather(executor, time.Now(), config)
+			_, err := gatherer.Gather(executor, time.Now(), config)
+			if err != nil {
+				t.Errorf("cant gather config %v", err)
```
This one is causing test failure. The error here is actually expected for some usecases, it shouldn't fail the test.

kushthedude replied:
So should I just mute the error over here?

wojtek-t replied:
Definiely you can't fail on error.
Either muting or introducing a per test-case parameter whether an error is expected or not is the way to go.

wojtek-t replied:
Removing return doesn't fix the problem - Error will still fire.

kushthedude replied:
So I will just log the error description in other format

kushthedude commented:
@wojtek-t Please review, I have just logged that error rather than failing the test at gatherer function.

wojtek-t commented on clusterloader2/pkg/measurement/common/slos/api_responsiveness_prometheus_test.go:
```diff
@@ -565,7 +573,10 @@ func TestLogging(t *testing.T) {
-			gatherer.Gather(executor, time.Now(), config)
+			_, err := gatherer.Gather(executor, time.Now(), config)
+			if err != nil {
+				t.Logf("cant gather config %v", err)
```
err doesn't mean we couldn't gather config. It may mean violation error.

Actually looking into we can do much better, i.e. check if this is violation error and fail if not:
https://github.com/kubernetes/perf-tests/blob/345ac7aeda560a246de34d5ce89419b7eca22a05/clusterloader2/pkg/errors/metric_violation_error.go#L37

So please change to:
```
if err != nil && !errors.IsMetricViolation(err) {
  t.Errorf("error while gathering results: %v", err)
}
```

wojtek-t replied:
[Also please squash all commits]

kushthedude replied:
done!

wojtek-t commented:
@kushthedude - please fix the compile errors and then LGTM

kushthedude commented:
> @kushthedude - please fix the compile errors and then LGTM

should be fixed now !

wojtek-t commented:
Thanks a lot for pushing this!

/lgtm
/approve

kushthedude commented:
Thanks @wojtek-t for bearing with me in this PR 😄
